builtin/logical/aws/secret_access_keys_test.go

```diff
@@ -47,53 +47,61 @@ func TestNormalizeDisplayName_NormNotRequired(t *testing.T) {
 }
 
 func TestGenUsername(t *testing.T) {
-
-	testUsername, warning, err := genUsername("name1", "policy1", "iam_user", `{{ printf "vault-%s-%s-%s-%s" (.DisplayName) (.PolicyName) (unix_time) (random 20) | truncate 64 }}`)
-	if err != nil {
-		t.Fatalf(
-			"expected no err; got %s",
-			err,
-		)
-	}
-
-	expectedUsernameRegex := `^vault-name1-policy1-[0-9]+-[a-zA-Z0-9]+`
-	require.Regexp(t, expectedUsernameRegex, testUsername)
-	// IAM usernames are capped at 64 characters
-	if len(testUsername) > 64 {
-		t.Fatalf(
-			"expected IAM username to be of length 64, got %d",
-			len(testUsername),
-		)
-	}
-
-	testUsername, warning, err = genUsername(
-		"this---is---a---very---long---name",
-		"long------policy------name",
-		"iam_user",
-		`{{ printf "%s-%s-%s-%s" (.DisplayName) (.PolicyName) (unix_time) (random 20) }}`,
-	)
-
-	if warning == "" || !strings.Contains(warning, "calling token display name/IAM policy name were truncated to 64 characters") {
-		t.Fatalf("expected a truncate warning; received empty string")
-	}
-	if len(testUsername) != 64 {
-		t.Fatalf("expected a username cap at 64 chars; got length: %d", len(testUsername))
-	}
-
-	testUsername, warning, err = genUsername("name1", "policy1", "sts", defaultSTSTemplate)
-	if strings.Contains(testUsername, "name1") || strings.Contains(testUsername, "policy1") {
-		t.Fatalf(
-			"expected sts username to not contain display name or policy name; got %s",
-			testUsername,
-		)
-	}
-	// STS usernames are capped at 64 characters
-	if len(testUsername) > 32 {
-		t.Fatalf(
-			"expected sts username to be under 32 chars; got %s of length %d",
-			testUsername,
-			len(testUsername),
-		)
+	type testCase struct {
+		name             string
+		policy           string
+		userType         string
+		UsernameTemplate string
+		expectedError    string
+		expectedRegex    string
+		expectedLength   int
+	}
+
+	tests := map[string]testCase{
```

Comment:
Can we also add an error case for custom template that generates a long username for the `sts` credential type since the cap in that case will be 32 chars?

Reply:
Added!